src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -697,6 +703,24 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
             if let Res::PrimTy(_) = res {
                 item.attrs.links.push((ori_link, None, fragment));
             } else {
+                // ~~WRONG: TODO: I think this happens too late and we need to instead put this in `self.resolve`~~
+                debug!("item {:?} resolved to {:?}", item, res);
+                if let Some(def_id) = res.opt_def_id() {
+                    let (link_vis, _span, _desc) = ::rustc_privacy::def_id_visibility(self.cx.tcx, def_id);
+                    if !matches!(link_vis, rustc_middle::ty::Visibility::Public) {
```

Comment:
Can't you just use `!=`?

Reply:
I ended up using a different approach that didn't involve messing around with `rustc_privacy` internals.